File: problem-005/solution.py
def cons(a, b):
    """
    Given two parameters, return a function that accepts a function and run it with previous parameters
    """
    def pair(f):
        return f(a, b)
    return pair


def car(pair):
    """
    Return first parameter.
    """
    return pair(lambda a, b: a)


def cdr(pair):
    """
    Return second parameter.
    """
    return pair(lambda a, b: b)


assert car(cons(3, 4)) == 3
assert cdr(cons(3, 4)) == 4

File: problem-005/solution.go
package problem005

type pairFunc func(x, y int) int
type pairFuncExecutor func(f pairFunc) int

func cons(a, b int) pairFuncExecutor {
	pair := func(f pairFunc) int {
		return f(a, b)
	}
	return pair
}

func car(f pairFuncExecutor) int {
	return f(func(x, y int) int {
		return x
	})
}

func cdr(f pairFuncExecutor) int {
	return f(func(x, y int) int {
		return y
	})
}
